src/main/java/org/sonar/plugins/stash/issue/collector/SonarQubeCollector.java

```diff
@@ -57,4 +72,87 @@ public static SonarQubeIssuesReport extractIssueReport(ProjectIssues projectIssu
 
     return result;
   }
+  
+  /**
+   * Extract Code Coverage report to be published into the pull-request.
+   * @throws SonarQubeClientException 
+   */
+  public static CoverageIssuesReport extractCoverageReport(String sonarQubeProjectKey, SensorContext context,
+      InputFileCacheSensor inputFileCacheSensor, String codeCoverageSeverity, SonarQubeClient sonarqubeClient) throws SonarQubeClientException {
+    
+    CoverageIssuesReport result = new CoverageIssuesReport();
+    
+    FileSystem fileSystem = inputFileCacheSensor.getFileSystem();
+    for (org.sonar.api.batch.fs.InputFile f : fileSystem.inputFiles(fileSystem.predicates().all())) {
+      
+      Double linesToCover = null;
+      Double uncoveredLines = null;
+
+      Measure<Integer> linesToCoverMeasure = context.getMeasure(context.getResource(f), CoreMetrics.LINES_TO_COVER);
+      if (linesToCoverMeasure != null) {
+        linesToCover = linesToCoverMeasure.getValue();
+      }
+      
+      Measure<Integer> uncoveredLinesMeasure = context.getMeasure(context.getResource(f), CoreMetrics.UNCOVERED_LINES);
+      if (uncoveredLinesMeasure != null) {
+        uncoveredLines = uncoveredLinesMeasure.getValue();
+      }
+      
+      // get lines_to_cover, uncovered_lines
+      if ((linesToCover != null) && (uncoveredLines != null)) {
+        double previousCoverage = sonarqubeClient.getCoveragePerFile(sonarQubeProjectKey, f.relativePath());
+  
+        CoverageIssue issue = new CoverageIssue(codeCoverageSeverity, f.relativePath());
+        issue.setLinesToCover(linesToCover);
+        issue.setUncoveredLines(uncoveredLines);
+        issue.setPreviousCoverage(previousCoverage);
+  
+        result.add(issue);
+        LOGGER.debug(issue.getMessage());
+      }
+    }
+  
+    // set previous project coverage from SonarQube server
+    double previousProjectCoverage = sonarqubeClient.getCoveragePerProject(sonarQubeProjectKey);
+    result.setPreviousProjectCoverage(previousProjectCoverage);
+
+    return result;
+  }
+  
+  
+  /**
+   * Extract Code Coverage retrieve by the SonarQube analysis.
+   */
+  public static double extractCoverage(String jsonBody) throws SonarQubeReportExtractionException {
+    double result = 0;
+    
+    try {
+      JSONArray jsonFiles = (JSONArray) new JSONParser().parse(jsonBody);
+      if (jsonFiles != null) {
+        
+        for (Object objectFile : jsonFiles.toArray()) {
+          JSONObject jsonFile = (JSONObject) objectFile;
+          
+          JSONArray jsonMeasures = (JSONArray) jsonFile.get("msr");
+          if (jsonMeasures != null) {

```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS134)